aws-s3/authorization.ml

```diff
@@ -103,6 +103,50 @@ let make_auth_header ~credentials ~scope ~signed_headers ~signature =
     signed_headers
     signature
 
+let make_presigned_url ?(scheme="https") ~credentials ~date ~region ~path ~bucket ~verb ~duration () =
+  let service = "s3" in
+  let ((y, m, d), ((h, mi, s), _)) = Ptime.to_date_time date in
+  let date = sprintf "%02d%02d%02d" y m d in
+  let time = sprintf "%02d%02d%02d" h mi s in
+  let host = sprintf "%s.s3.amazonaws.com" bucket in
+  let duration = string_of_int duration in
+  let region = Region.to_string region in
+  let headers =
+      [ ("Host", host);
+      ]
+      |> List.fold_left ~f:(fun acc (key, value) -> Headers.add ~key ~value acc) ~init:Headers.empty
+  in
+  let query = [
+        ("X-Amz-Algorithm", "AWS4-HMAC-SHA256");
+        ("X-Amz-Credential", sprintf "%s/%s/%s/s3/aws4_request" credentials.Credentials.access_key date region);
+        ("X-Amz-Date", sprintf "%sT%sZ" date time);
+        ("X-Amz-Expires", duration);
+        ("X-Amz-SignedHeaders", "host");
+      ] in
+  let scope = make_scope ~date ~region ~service in
+  let signing_key = make_signing_key ~date ~region ~service ~credentials () in
+  let signature, _signed_headers =
+    make_signature ~date ~time ~verb ~path ~headers ~query ~signing_key ~scope ~payload_sha:"UNSIGNED-PAYLOAD"
+  in
+  let base_uri = Uri.make ~scheme ~host ~path () in
+  List.fold_left ~f:Uri.add_query_param'~init:base_uri (List.append query [("X-Amz-Signature", signature)] |> List.rev)
```

Comment:
Why not add the query params on the Uri.make function invocation (~query)  - Or do you need fine grained control over the query param order? 

```ocaml
let query = 
  ("X-Amz-Signature", signature) :: query 
  |> List.map ~f:(fun (k, v) -> (k, [v]))
in
Uri.make ~scheme ~host ~path ~query ()
```